### PR TITLE
ENH: Create framework for registering new modules 

### PR DESCRIPTION
**Pull request by mohammadrashid0917**
registerModule function

Notes:
-most of the code is written in qMRMLVRHomeWidget because I always needed to access the buttons of the widget, which are all members of qMRMLVRHomeWidgetPrivate, and I can't access that from qMRMLVirtualRealityView.cxx 

-I couldn't make the QMap that stores the pushButtons and Widgets a member of qMRMLVRHomeWidgetPrivate (I could make it, but I couldn't use it), I tried a lot of different things and couldn't avoid an error, I think it has to do with D-pointers and the const keyword. 

-To hide the widget when the back button is pressed, I used a for loop and an iterator object to just set visibility of every widget in the map as false. This seems like a brute force way to do it and there may be a better way (the only other way I could think of is having yet another map for the back button)

Also, with each one of my pull requests, all of my previous commits are also being included even though you already merged those ones...not sure why that is or how I can prevent it.

### CODE REVIEW

**Comment by cpinter** [comment]
Q_D(const q....Widget)

Please fix this and redo PR. Thanks!

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 1]
```diff
@@ -621,13 +632,22 @@ qMRMLVirtualRealityView::qMRMLVirtualRealityView(QWidget* _parent) : Superclass(
...
+
 }
 
 // --------------------------------------------------------------------------
 qMRMLVirtualRealityView::~qMRMLVirtualRealityView()
 {
 }
 
+//------------------------------------------------------------------------------
+void qMRMLVirtualRealityView::registerModule(QWidget* widget, QIcon* icon) {
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: Braces always in new line please

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 2]
```diff
@@ -621,13 +632,22 @@ qMRMLVirtualRealityView::qMRMLVirtualRealityView(QWidget* _parent) : Superclass(
...
+  Q_D(qMRMLVirtualRealityView);
+
+  d->HomeWidget->setupModuleButton(widget, icon);
+
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: Please remove this empty line and any other empty line that has no purpose (e.g. separate two blocks of code)

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 3]
```diff
@@ -106,6 +113,55 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
   return d->VirtualRealityViewNode;
 }
 
+//-----------------------------------------------------------------------------
+void qMRMLVirtualRealityHomeWidget::setupModuleButton(QWidget* widget, QIcon* icon)
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Rename function to addModuleButton

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 4]
```diff
@@ -106,6 +113,55 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+ 
+  moduleButton->setVisible(true);
+  widget->setVisible(false);
+  d->backButton->setVisible(false);
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLVirtualRealityHomeWidget::onModuleButtonPressed()
+{
+  Q_D(const qMRMLVirtualRealityHomeWidget);
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: No need for const here because the function is not a const.

You would only need the const if it was declared like this:
void onModuleButtonPressed()const;

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 5]
```diff
@@ -106,6 +113,55 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+//-----------------------------------------------------------------------------
+void  qMRMLVirtualRealityHomeWidget::onBackButtonPressed()
+{
+  Q_D(const qMRMLVirtualRealityHomeWidget);
+  d->HomeWidgetFrame->setVisible(true);
+  d->ModulesGroupBox->setVisible(true); 
+  d->backButton->setVisible(false);
+
+  QMap<QPushButton*, QWidget*>& map = this->ModuleWidgetsMap;
+  QMap<QPushButton*, QWidget*>::const_iterator i;
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Never use variable names like "map" and "i". It's OK if your for loop declaration is long.

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 6]
```diff
@@ -106,6 +113,55 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+void  qMRMLVirtualRealityHomeWidget::onBackButtonPressed()
+{
+  Q_D(const qMRMLVirtualRealityHomeWidget);
+  d->HomeWidgetFrame->setVisible(true);
+  d->ModulesGroupBox->setVisible(true); 
+  d->backButton->setVisible(false);
+
+  QMap<QPushButton*, QWidget*>& map = this->ModuleWidgetsMap;
+  QMap<QPushButton*, QWidget*>::const_iterator i; 
+  for (i = map.constBegin(); i != map.constEnd(); ++i) {
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Braces always in new lines please. Read the style guide again if you feel you need it.

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 7]
```diff
@@ -106,6 +113,55 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+    i.value()->setVisible(false);
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: I think this is fine, although it looks brute force. If you set visible to false when it's already false then nothing happens in Qt, so you're not adding extra computation here. I think if you added cache variables it would be actually less clean.

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.h [hunk 8]
```diff
@@ -66,16 +70,24 @@ public slots:
...
+
 protected:
   QScopedPointer<qMRMLVirtualRealityHomeWidgetPrivate> d_ptr;
+  
 
 private:
   Q_DECLARE_PRIVATE(qMRMLVirtualRealityHomeWidget);
   Q_DISABLE_COPY(qMRMLVirtualRealityHomeWidget);
+
+  QMap<QPushButton*, QWidget*> ModuleWidgetsMap;
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.h]: I think this should go into the private class. Remember, anything that the users of the public class shou;dn't have access to should be in the private class (the map is private anyway, and the idea of the private class is to minimize changes in the header file). If we need access to the map, we should add functions for that later.

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 9]
```diff
@@ -93,8 +96,12 @@ qMRMLVirtualRealityViewPrivate::qMRMLVirtualRealityViewPrivate(qMRMLVirtualReali
   , CamerasLogic(NULL)
 {
   this->MRMLVirtualRealityViewNode = 0;
+  this->HomeWidget = new qMRMLVirtualRealityHomeWidget(q_ptr);
+
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: You're adding too much unnecessary whitespace that makes the code harder to read

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 10]
```diff
@@ -119,6 +126,10 @@ CTK_GET_CPP(qMRMLVirtualRealityView, vtkOpenVRRenderWindow*, renderWindow, Rende
 //----------------------------------------------------------------------------
 CTK_GET_CPP(qMRMLVirtualRealityView, vtkOpenVRRenderWindowInteractor*, interactor, Interactor);
 
+//---------------------------------------------------------------------------
+
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: Three empty lines here. Why...

**Comment by mohammadrashid0917** [comment]
Sorry ignore that push, I didn't amend correctly

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 11]
```diff
@@ -106,6 +113,57 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
   return d->VirtualRealityViewNode;
 }
 
+//-----------------------------------------------------------------------------
+void qMRMLVirtualRealityHomeWidget::addModuleButton(QWidget* widget, QIcon* icon)
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Since you need to pass QIcon as not a pointer but as value, please change the signature of this function to
qMRMLVirtualRealityHomeWidget::addModuleButton(QWidget* widget, QIcon& icon)
(& instead of *)
This way your program won't crash if icon is nullptr.
That said please add a check for
if (!widget)
{
  qCritical().....
  return
}
Please try not to forget checking the inputs. Otherwise your program will crash!

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 12]
```diff
@@ -106,6 +113,57 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+  moduleButton->setIcon(*icon);
+  
+  widget->setParent(d->ModuleWidgetFrame);
+  d->ModuleWidgetsMap[moduleButton] = widget;
+
+  QObject::connect(moduleButton, SIGNAL(clicked()), this, SLOT(onModuleButtonPressed()));
+  QObject::connect(d->backButton, SIGNAL(clicked()), this, SLOT(onBackButtonPressed()));
+ 
+  moduleButton->setVisible(true);
+  widget->setVisible(false);
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Please rename widget to moduleWidget everywhere (including the function declaration in header) for clarity

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 13]
```diff
@@ -106,6 +113,57 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+  d->backButton->setVisible(false);
+}
+
+//-----------------------------------------------------------------------------
+void qMRMLVirtualRealityHomeWidget::onModuleButtonPressed()
+{
+  Q_D(qMRMLVirtualRealityHomeWidget);
+ 
+  QPushButton* moduleButton = qobject_cast<QPushButton*>(sender());
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Check if nullptr

This is important stuff!

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 14]
```diff
@@ -106,6 +113,57 @@ vtkMRMLVirtualRealityViewNode* qMRMLVirtualRealityHomeWidget::virtualRealityView
...
+//-----------------------------------------------------------------------------
+void  qMRMLVirtualRealityHomeWidget::onBackButtonPressed()
+{
+  Q_D(qMRMLVirtualRealityHomeWidget);
+  d->HomeWidgetFrame->setVisible(true);
+  d->ModulesGroupBox->setVisible(true); 
+  d->backButton->setVisible(false);
+
+  QMap<QPushButton*, QWidget*>::const_iterator iteratorForMap; 
+  for (iteratorForMap = d->ModuleWidgetsMap.constBegin();
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: We're not using 80 character terminals any more. Just let the for declaration go in one line. It will be easier to read. Thanks!

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx [hunk 15]
```diff
@@ -160,6 +218,7 @@ void qMRMLVirtualRealityHomeWidget::updateWidgetFromMRML()
   d->Magnification1xButton->setEnabled(vrViewNode != nullptr && vrViewNode->GetMagnification() != NULL);
   d->Magnification10xButton->setEnabled(vrViewNode != nullptr && vrViewNode->GetMagnification() != NULL);
   d->Magnification100xButton->setEnabled(vrViewNode != nullptr && vrViewNode->GetMagnification() != NULL);
+
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.cxx]: Remove blank line

File: VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.h [hunk 16]
```diff
@@ -66,13 +70,17 @@ public slots:
...
+  void onBackButtonPressed();
+
 protected slots:
   /// Update widgets from the MRML node
   void updateWidgetFromMRML();
 
 protected:
   QScopedPointer<qMRMLVirtualRealityHomeWidgetPrivate> d_ptr;
-
+
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityHomeWidget.h]: Remove spaces in line (the Slicer precommit hooks would make your commit fail if something like this was there, so it's good to pay attention).
To clarify, I don't want you to remove the line. The problem is the "trailing whitespaces"

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 17]
```diff
@@ -57,7 +57,10 @@
 #include "vtkSlicerConfigure.h" // For Slicer_USE_OpenVR
 #include "vtkSlicerCamerasModuleLogic.h"
 
-// VirtualReality includes
+//VirtualReality Widgets includes
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: Space after //
Consistency, attention to detail, code readability...

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 18]
```diff
@@ -93,6 +96,7 @@ qMRMLVirtualRealityViewPrivate::qMRMLVirtualRealityViewPrivate(qMRMLVirtualReali
   , CamerasLogic(NULL)
 {
   this->MRMLVirtualRealityViewNode = 0;
+  this->HomeWidget = new qMRMLVirtualRealityHomeWidget(q_ptr);
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx:96]: How do you access this one? I cannot see a getter function in the view.

File: VirtualReality/Widgets/qMRMLVirtualRealityView_p.h [hunk 19]
```diff
@@ -40,6 +40,9 @@
 #include <ctkPimpl.h>
 #include <ctkVTKObject.h>
 
+//VirtualReality Widgets includes
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView_p.h]: Space after //

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx [hunk 20]
```diff
@@ -57,7 +57,10 @@
 #include "vtkSlicerConfigure.h" // For Slicer_USE_OpenVR
 #include "vtkSlicerCamerasModuleLogic.h"
 
-// VirtualReality includes
+//VirtualReality Widgets includes
+#include "qMRMLVirtualRealityHomeWidget.h"
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.cxx]: You won't need this include here, as it's included in the private header

File: VirtualReality/qSlicerHomeVirtualWidget.cxx [hunk 21]
```diff
@@ -1,93 +0,0 @@
-/*==============================================================================
```
Comment (cpinter) [inline VirtualReality/qSlicerHomeVirtualWidget.cxx]: Did you actually delete the widget???

File: VirtualReality/Widgets/qMRMLVirtualRealityView.cxx (hunk 18 above)
Comment (mohammadrashid0917) [reply VirtualReality/Widgets/qMRMLVirtualRealityView.cxx:96]: That line just initializes the member variable HomeWidget, then I access the member directly through a D-pointer in the registerModule function. Should I make a getter function and access it through that instead?

File: VirtualReality/qSlicerHomeVirtualWidget.cxx (hunk 21 above)
Comment (mohammadrashid0917) [reply VirtualReality/qSlicerHomeVirtualWidget.cxx]: Not sure how this happened, I don't remember deleting these files but they indeed aren't in my /VirtualReality/ directory. All I can think of is that we mistakenly deleted these when when you came by on Monday. Anyways I've taken them from your SlicerVR fork and put them back

File: VirtualReality/Widgets/qMRMLVirtualRealityView.h [hunk 22]
```diff
@@ -100,6 +101,9 @@ class Q_SLICER_QTMODULES_VIRTUALREALITY_WIDGETS_EXPORT qMRMLVirtualRealityView :
   /// Get underlying RenderWindow
   Q_INVOKABLE bool isHardwareConnected()const;
 
+  void registerModule(QWidget* widget, QIcon& icon);
+  qMRMLVirtualRealityHomeWidget* vrHomeWidget();
```
Comment (cpinter) [inline VirtualReality/Widgets/qMRMLVirtualRealityView.h]: Make this const please

**Comment by cpinter** [comment]
Integrated in https://github.com/cpinter/SlicerVirtualReality/commit/24fb665fbc76254e0644ab4c165ff7ceeda4457d
